Meta/CMake/serenity_toolchain.cmake

```diff
@@ -0,0 +1,30 @@
+#
+# Define and generate CMake toolchain to compile serenity with
+#
+
+set(SERENITY_ARCH "i686" CACHE STRING "Target architecture for SerenityOS (i686 or x86_64)")
+if("${SERENITY_ARCH}" STREQUAL "i686")
+    set(SERENITY_CLANG_ARCH "i386")
+else()
+    set(SERENITY_CLANG_ARCH ${SERENITY_ARCH})
+endif()
+
+if (${CMAKE_HOST_SYSTEM_NAME} MATCHES SerenityOS)
+    message("Good job on building cmake!")
+else()
+    # Note: We must generate the toolchain before calling project, as project looks for the toolchain entries
+    set(SERENITY_TOOLCHAIN "GNU" CACHE STRING "Compiler toolchain to use for Serenity (GNU or Clang)")
+    set(SERENITY_BUILD_DIR "${CMAKE_CURRENT_BINARY_DIR}")
+    set(SERENITY_SOURCE_DIR "${CMAKE_SOURCE_DIR}")
```

Comment:
It's `SerenityOS_SOURCE_DIR` elsewhere, but `SERENITY_SOURCE_DIR` here? don't think that was something you introduced, but maybe since you are cleaning this all up, this could get fixed too?

Reply:
I tried a few approaches for the toolchain, including the "normal" way which seems to be environment varaibles? But doing that required setting SERENITY_SOURCE_DIR and SERENITY_ARCH every single time you did a rebuild, which is not very nice.

The only place these two variables are used is when generating the Toolchain file, like so:
```
set(triple @SERENITY_ARCH@-pc-serenity)
set(TOOLCHAIN_PATH @SERENITY_SOURCE_DIR@/Toolchain/Local/@SERENITY_ARCH@/bin)
set(TOOLCHAIN_PREFIX ${TOOLCHAIN_PATH}/${triple}-)

# where to read from/write to
set(CMAKE_SYSROOT @SERENITY_BUILD_DIR@/Root)
```
Up for bike shedding these guys though.

To answer the actual question, SerenityOS_SOURCE_DIR is a variable that gets set by he `project(SerenityOS)` call, and this toolchain cmake helper must be include()'d *before* that so that we can set CMAKE_TOOLCHAIN_FILE to the newly autogenerated toolchain file in the Build directory. The project() call is the place in the CMakeLists.txt that CMake detects what compiler/OS/etc we're compiling for, and where it reads CMAKE_TOOLCHAIN_FILE.

Reply:
That makes sense once explained, thanks!